server/worker/tasks.py

```diff
@@ -133,22 +149,60 @@ def run_task(task: BackgroundTask, db_session):
 
 
 def claim_next_task(worker_id: str, db_session) -> Optional[BackgroundTask]:
+    # Use SELECT ... FOR UPDATE to lock all queued tasks, ensuring that only a
+    # single worker can claim a single task at a time. This ensures that we get
+    # an accurate view of which tasks are available to claim. If another worker has
+    # already locked the queued task, this query will block (though it shouldn't
+    # take long to claim a task, so this shouldn't cause problems).
+    #
+    # If new tasks are added after we lock the queued tasks, they won't be
+    # accessible by this worker, but another worker also won't be able to claim
+    # them (since it will also be trying to lock _all_ queued tasks, including
+    # the ones we already locked).
+    #
+    # Importantly, this also ensures that we get a safe view of which lock_keys
+    # are currently in use by in-progress tasks. If we're the only worker that
+    # can claim a task right now, then no additional lock_keys can be locked. It
+    # is possible for another worker to finish an in-progress task and release a
+    # lock_key while we're running this query, but that's fine because even if
+    # we don't see that change and wanted to claim that lock_key, we'll just fail
+    # and try again later.
+    #
+    # Another approach to this would be to lock the entire table, but that would
+    # also prevent in-progress tasks from completing.
+    #
+    # We don't actually need to load the queued tasks into the app, so we just
+    # execute the query.
+    (
+        db_session.execute(
+            db_session.query(BackgroundTask)
+            .filter_by(started_at=None)
+            .with_for_update()
+            .with_entities(BackgroundTask.id)
+        )
+    )
+    in_progress_task_lock_keys = (
+        db_session.query(BackgroundTask)
+        .filter(BackgroundTask.started_at.isnot(None))
+        .filter_by(completed_at=None)
+        .with_entities(BackgroundTask.lock_key)
+        .subquery()
+    )
```

Comment:
Even if we've reasoned through the fact that this system works and covers all edge cases, I am wondering if it's still preferable to just lock the entire tasks table when we look for a task.

I'm finding myself re-reading the above explanatory comment and code a few times to understand the mechanism.

And as noted elsewhere in this PR, it seems like there isn't really much downside to locking the entire tasks table. Yes, task completion might be blocked for just a moment, but that should be inconsequential relative to overall task processing time. And the mental model really is much simpler.

I'm ultimately okay with the current approach but just wanted to flag!

Reply:
I had the same thought yesterday during that discussion. I experimented with locking the entire tasks table but, surprisingly, I couldn't get it working. All workers ended up hanging when trying to lock the table. There's likely some explanation for this, but I'm not sure whether it's worth spending more time debugging to try to figure out why.

Do you think it's worth spending some more time on it? Or just going with what I have in the PR already (which seems to be working)?

Reply:
Oh interesting 🤔 Yeah in that case, I'm fine with merging this to have something working in place, so long as we're confident in the approach and not concerned about tricky to debug issues during audit season itself (which I believe is the case!)